File: leetCode/binaryTree/Same_Tree_100/Main.java
import java.util.LinkedList;
import java.util.Queue;

import javax.swing.tree.TreeNode;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSameTreeRecursive(TreeNode p, TreeNode q) {
        if(p == null && q == null) return true;
        if(p == null || q == null) return false;
        return (p.val == q.val) && isSameTree(p.left, q.left) && isSameTree(p.right, q.right); 
    }

    public boolean isSameTree(TreeNode p, TreeNode q) {
        Queue<TreeNode> nodes = new LinkedList<>();
        nodes.offer(p);
        nodes.offer(q);
        while(!nodes.isEmpty()) {
            TreeNode n1 = nodes.poll();
            TreeNode n2 = nodes.poll();
            if(n1 == null && n2 == null) continue;
            if(n1 == null || n2 == null) return false;
            if(n1.val != n2.val) return false;
            nodes.offer(n1.left);
            nodes.offer(n2.left);
            nodes.offer(n1.right);
            nodes.offer(n2.right);
        }
        return true;
    }
}

File: leetCode/binaryTree/Same_Tree_100/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque


class Solution:
    def isSameTree_rec(self, p: TreeNode, q: TreeNode) -> bool:
        if not p and not q:
            return True
        if not p or not q:
            return False
        return p.val == q.val and self.isSameTree_rec(p.left, q.left) and self.isSameTree_rec(p.right, q.right)

    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        nodes = deque()
        if p and q:
            nodes.append(p)
            nodes.append(q)
        elif p or q:
            return False

        while nodes:
            curr_p = nodes.popleft()
            curr_q = nodes.popleft()

            if curr_p.val != curr_q.val:
                return False

            if curr_p.left and curr_q.left:
                nodes.append(curr_p.left)
                nodes.append(curr_q.left)
            elif curr_p.left or curr_q.left:
                return False

            if curr_p.right and curr_q.right:
                nodes.append(curr_p.right)
                nodes.append(curr_q.right)
            elif curr_p.right or curr_q.right:
                return False

        return True
